lib/basedriver/commands/execute.js

```diff
@@ -0,0 +1,102 @@
+import _ from 'lodash';
+import log from '../logger';
+import vm from 'vm';
+import { attach } from 'webdriverio';
+import { MJSONWP_ELEMENT_KEY, W3C_ELEMENT_KEY } from '../../protocol/protocol';
+
+const BASE_PATH = '/wd/hub'; // TODO defining this here is brittle but it is hardcoded in routes, so...
+
+let commands = {};
+
+/*
+ * TODO fill out entire docstring
+ *
+ * This method takes a string which is executed as javascript in the context of
+ * a new nodejs VM, and which has available a webdriverio driver object, having
+ * already been attached to the currently running session.
+ */
+commands.executeDriverScript = async function (script, scriptType) {
+  if (!!scriptType && scriptType !== 'webdriverio') {
+    throw new Error("Only the 'webdriverio' script type is currently supported");
+    // TODO should probably add support for WD script type as well?
+  }
+
+  if (!this.opts.host || !this.opts.port) {
+    throw new Error('Host or port of running server were not defined; this is required');
+  }
+  const driver = attach({
+    sessionId: this.sessionId,
+    protocol: 'http', // Appium won't ever be behind ssl locally
+    hostname: this.opts.host,
+    port: this.opts.port,
+    path: BASE_PATH,
+    isW3C: true, // TODO should probably retrieve client protocol and declare the same one for the script
+    capabilities: this.caps
+  });
+  const fullScript = buildScript(script);
+  const result = await vm.runInNewContext(fullScript)(driver);
```

Comment:
Do we have access to the server's file system from this script? If yes then I wouldn't appreciate this approach much

Reply:
how are we going to show errors caused by script interpretation?

Reply:
how can we guarantee that this piece of code won't loop forever like it is described in https://nodejs.org/api/vm.html#vm_timeout_limitations_when_using_process_nexttick_promises_and_queuemicrotask?

Reply:
We can add a timeout as a parameter for the vm. And code run in the vm doesn't have access to the filesystem since it has no access to `require`, so you can't e.g. `require('fs')`.

Errors caused by script interpretation are thrown and sent to the client. I should probably add a test for errors that cause the script not to even compile, but I bet it works with the code as is.